src/ray/common/id.cc

```diff
@@ -104,23 +173,60 @@ TaskID ObjectID::TaskId() const {
       std::string(reinterpret_cast<const char *>(id_), TaskID::Size()));
 }
 
-ObjectID ObjectID::ForPut(const TaskID &task_id, int64_t put_index) {
-  RAY_CHECK(put_index >= 1 && put_index <= kMaxTaskPuts) << "index=" << put_index;
+ObjectID ObjectID::ForPut(const TaskID &task_id, uint32_t put_index) {
+  RAY_CHECK(put_index >= 1 && put_index <= MAX_OBJECT_INDEX) << "index=" << put_index;
+
+  uint16_t flags = 0x0000;
+  object_id_helper::SetIsTaskFlag(&flags, true);
+  object_id_helper::SetObjectTypeFlag(&flags, ObjectType::PUT_OBJECT);
+  object_id_helper::SetTransportTypeFlag(&flags, TransportType::STANDARD);
+
   ObjectID object_id;
-  std::memcpy(object_id.id_, task_id.Binary().c_str(), task_id.Size());
-  object_id.index_ = -put_index;
+  std::memcpy(object_id.id_, task_id.Binary().c_str(), TaskID::LENGTH);
+  std::memcpy(object_id.id_ + TaskID::LENGTH, &flags, sizeof(flags));
+  std::memcpy(object_id.id_ + TaskID::LENGTH + FLAGS_BYTES_LENGTH, &put_index, sizeof(put_index));
   return object_id;
 }
 
-ObjectID ObjectID::ForTaskReturn(const TaskID &task_id, int64_t return_index) {
-  RAY_CHECK(return_index >= 1 && return_index <= kMaxTaskReturns)
+uint32_t ObjectID::ObjectIndex() const {
+  uint32_t index;
+  std::memcpy(&index, id_ + TaskID::LENGTH + FLAGS_BYTES_LENGTH, sizeof(index));
+  return index;
+}
+
+ObjectID ObjectID::ForTaskReturn(const TaskID &task_id, uint32_t return_index, TransportType transport_type) {
```

Comment:
The `ObjectID` index also seems worth `typedef`ing, instead of using a raw `uint32_t`.